### PR TITLE
Graphite WIP

### PR DESCRIPTION
**Pull request by corylanou**
Graphite end point has been beaten back into shape.  Added test coverage.

Changed config to allow for infinite Graphite listeners... might be something we need to limit at some point.

would love feedback!

### CODE REVIEW

File: graphite/graphite.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	TCPAddr *net.TCPAddr
+
+	// The UDP address to listen on.
+	UDPAddr *net.UDPAddr
+
+	// The name of the database to insert data into.
+	Database string
+}
+
+// ListenAndServe opens TCP (and optionally a UDP) socket to listen for messages.
```
Comment (corylanou) [inline graphite/graphite.go]: They can specify one or the other (tcp or udp).  So they can just open a udp if in the config they only specify the udp protocol.

Comment (otoolep) [inline graphite/graphite.go]: Agreed. It was just the comment doesn't mention anything about TCP being optional.

File: graphite/graphite.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrDatabaseNotSpecified retuned when no database was specified in the config file
+	ErrDatabaseNotSpecified = errors.New("database was not specified in config")
+
+	// ErrServerNotSpecified returned when Server is not specified.
+	ErrServerNotSpecified = errors.New("server not present")
+)
```
Comment (pauldix) [inline graphite/graphite.go]: We've been putting in `errors.go`. @benbjohnson you think it makes sense to put these in there too or keep them here?

File: graphite/graphite.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+		m.Tags,
+		m.Timestamp,
+		values,
+	); err != nil {
+		return fmt.Errorf("write series data: %s", err)
+	}
+	return nil
+}
+
+type Metric struct {
```
Comment (pauldix) [inline graphite/graphite.go]: Should `Metric` be an exported type? If so I think we should have some standardization on this between the HTTP API, Graphite, and other input plugins.

File: graphite/graphite.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+}
+
+func DecodeNameAndTags(field string) (string, map[string]string, error) {
+	var (
+		name string
+		tags = make(map[string]string)
+	)
+
+	// decode the name and tags
+	values := strings.Split(field, `.`)
```
Comment (pauldix) [inline graphite/graphite.go]: I'd like the separator to be something that is configurable via the config file. `.` should be the default, but if they're using `-` or `_` they should be able to do that too.

File: graphite/graphite.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+	if len(values)%2 != 1 {
+		// There should always be an odd number of fields to map a metric name and tags
+		// ex: region.us-west.hostname.server01.cpu -> tags -> region: us-west, hostname: server01, metric name -> cpu
+		return name, tags, fmt.Errorf("received %q which doesn't conform to format of key.value.key.value.metric or metric", field)
+	}
+
+	// Name is the last field
+	name = values[len(values)-1]
```
Comment (pauldix) [inline graphite/graphite.go]: They should be able to set `name` as either the first or last field via the config.

**Comment by pauldix** [comment]
@corylanou: just a few minor comments, but overall looks good.

File: graphite/graphite.go (hunk 2 above)
Comment (benbjohnson) [inline graphite/graphite.go]: The `graphite` package is pretty small. I'd leave them here.

File: graphite/graphite.go (hunk 5 above)
Comment (corylanou) [inline graphite/graphite.go]: @pauldix for clarification, all graphite logs in this format:

metric_path value timestamp\n

are you asking to be able to do for the metric_path:

name.seriesName.seriesValue -> cpu.host.server01

as well as 

seriesName.seriesValue.name -> host.server01.cpu

If so, in the config would we just want
name: first

or 

name: last

Thoughts?

File: cmd/influxd/config.go [hunk 6]
```diff
@@ -251,6 +248,71 @@ func ParseConfig(s string) (*Config, error) {
...
 }
 
+type Graphite struct {
+	Address       string `toml:"address"`
+	Database      string `toml:"database"`
+	Enabled       bool   `toml:"enabled"`
+	Port          int    `toml:"port"`
+	Protocol      string `toml:"protocol"`
+	NamePosition  string `toml:"name-position"`
+	NameSeperator string `toml:"name-seperator"`
```
Comment (otoolep) [inline cmd/influxd/config.go]: Nit: separator is misspelled.

File: cmd/influxd/config.go [hunk 7]
```diff
@@ -251,6 +248,71 @@ func ParseConfig(s string) (*Config, error) {
...
+}
+
+// Default carbon port per http://graphite.readthedocs.org/en/1.0/feeding-carbon.html
+const defaultGrahitePort = 2004
+
+// TCPAddr returns the TCP address to connect on.
+// If port is not specified in the config file, it will default to defaultGraphitePort (2004)
+// If address is not specified,it will default to the defaultBindAddress passed in
+func (g Graphite) TCPAddr(defaultBindAddress string) *net.TCPAddr {
```
Comment (otoolep) [inline cmd/influxd/config.go]: I know we've been beating on this function, so sorry if it's getting tiresome, but this implementation is a little inconsistent. I would either push or defaults in, or none. In other words, I think the function signature might be better if:
```
TCPAddr(defaultBind, defaultPort)
```
Not a big deal, just noticed it in passing.

File: graphite/graphite.go [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+	WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+	DefaultRetentionPolicy(database string) (*influxdb.RetentionPolicy, error)
+}
+
+func NewServer(is influxdbServer) *Server {
+	s := Server{Server: is}
+	return &s
+}
+
+// ListenAndServe opens TCP (and optionally a UDP) socket to listen for messages.
```
Comment (otoolep) [inline graphite/graphite.go]: Comment is not accurate.

**Comment by otoolep** [comment]
I realise we're under the time-pressure to get this code in. I would like to suggest a slightly different way to implement this, that could simplify the code and make it easier to maintain (and remove the need for some functions).

This line is key:
```
s := graphite.Server{Server: s}
```
I believe you should instead bring either a TCP Graphite Server into existence, or a UDP Graphite Server into existence, depending on the "protocol" flag, as you walk the config blocks. Each of these types would take and address *string* -- in the form of address:port (hence no need for `TCPAddr()` and `UDPAddr()`. Then you wouldn't need `TCPAddr()`, `UDPAddr()` -- all that stuff would be encapsulated inside the respective types. The two Server types would simply have `New...`, `Listen`, and maybe `Stop` call. You could even define an Interface that both Graphite Servers must adhere too.

It depends how much you want to change you implementation, but now that I see what you have done I think there is a little too much code outside of the Server, and not enough inside it (inside two types that need to be created -- TcpGraphite and UdpGraphite). Make sense?

**Comment by otoolep** [comment]
If you look at the syslog-gollector I wrote last year, it does more-or-less this. It listens on TDP and UDP ports, but most of the logic is inside the types:

https://github.com/otoolep/syslog-gollector/blob/master/syslog-gollector.go#L180
https://github.com/otoolep/syslog-gollector/blob/master/syslog-gollector.go#L190

(This was some of the earliest Go code I wrote, so it's pretty naive, but you get the idea).

**Comment by otoolep** [comment]
FWIW, this is really important code, since Graphite is a big deal. We could be in this code a lot, so I think it's worth taking the time to be sure you're happy with it.

**Comment by benbjohnson** [comment]
I'd agree with @otoolep. It seems like there's a lot of prefixing of `TDP` and `UDP` and maybe it's best if they're separate types. The actual parsing code and insertion code can be shared as separate functions but everything else would probably be cleaner if separated.

Also, the types should be `graphite.TCPServer` & `graphite.UDPServer` instead of using camel case.

https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Initialisms

It might be a good idea to merge this code as is so that the PR for the refactor to separate types will be easier to read.

**Comment by otoolep** [comment]
FWIW, once the build is green I am OK with merging and then refactoring. @corylanou and I are going to look at the code today, and if the change is overall simpler perhaps we'll just push up a new PR. If not, up for merging first.

**Comment by corylanou** [comment]
New pull request in process that replaces this one. https://github.com/influxdb/influxdb/pull/1294

**Pull request closed without merging**